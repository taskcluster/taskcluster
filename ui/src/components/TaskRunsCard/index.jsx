import React, { Fragment, Component } from 'react';
import { withRouter } from 'react-router-dom';
import classNames from 'classnames';
import { func, number, string } from 'prop-types';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import Label from '@mozilla-frontend-infra/components/Label';
import { withStyles } from '@material-ui/core/styles';
import { fade } from '@material-ui/core/styles/colorManipulator';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardContent from '@material-ui/core/CardContent';
import Collapse from '@material-ui/core/Collapse';
import MobileStepper from '@material-ui/core/MobileStepper';
import Divider from '@material-ui/core/Divider';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';
import ChevronLeftIcon from 'mdi-react/ChevronLeftIcon';
import ChevronUpIcon from 'mdi-react/ChevronUpIcon';
import ChevronDownIcon from 'mdi-react/ChevronDownIcon';
import ChevronRightIcon from 'mdi-react/ChevronRightIcon';
import ContentCopyIcon from 'mdi-react/ContentCopyIcon';
import LinkIcon from 'mdi-react/LinkIcon';
import LockIcon from 'mdi-react/LockIcon';
import LockOpenOutlineIcon from 'mdi-react/LockOpenOutlineIcon';
import OpenInNewIcon from 'mdi-react/OpenInNewIcon';
import Button from '../Button';
import ConnectionDataTable from '../ConnectionDataTable';
import DateDistance from '../DateDistance';
import StatusLabel from '../StatusLabel';
import NoRunsIcon from './NoRunsIcon';
import { ARTIFACTS_PAGE_SIZE } from '../../utils/constants';
import { runs } from '../../utils/prop-types';
import Link from '../../utils/Link';

const DOTS_VARIANT_LIMIT = 5;

@withRouter
@withStyles(
  theme => ({
    headline: {
      paddingLeft: theme.spacing.double,
      paddingRight: theme.spacing.double,
    },
    cardContent: {
      paddingLeft: 0,
      paddingRight: 0,
      paddingTop: theme.spacing.double,
      paddingBottom: theme.spacing.double,
      '&:last-child': {
        paddingBottom: theme.spacing.triple,
      },
    },
    controls: {
      display: 'flex',
      alignItems: 'center',
      paddingLeft: theme.spacing.unit,
      paddingBottom: theme.spacing.unit,
    },
    listItemButton: {
      ...theme.mixins.listItemButton,
    },
    pointer: {
      cursor: 'pointer',
    },
    logButton: {
      marginRight: theme.spacing.unit,
    },
    artifactsListItemContainer: {
      display: 'block',
    },
    boxVariantIcon: {
      width: '25%',
      height: 'auto',
    },
    boxVariant: {
      textAlign: 'center',
    },
    boxVariantText: {
      color: fade(theme.palette.text.primary, 0.4),
    },
    artifactLink: {
      textDecoration: 'none',
      display: 'flex',
      justifyContent: 'space-between',
      verticalAlign: 'middle',
    },
    artifactTableRow: {
      height: 'auto',
    },
    artifactNameWrapper: {
      display: 'inline-flex',
      flexBasis: '50%',
    },
    artifactName: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
    },
    liveLogLabel: {
      marginLeft: theme.spacing.unit / 2,
      marginBottom: theme.spacing.unit / 2,
    },
  }),
  { withTheme: true }
)
/**
 * Render a paginated card layout for the runs of a GraphQL task response.
 */
export default class TaskRunsCard extends Component {
  static propTypes = {
    /**
     * A collection of runs for a GraphQL task.
     */
    runs: runs.isRequired,
    /**
     * The worker type associated with the runs' task.
     */
    workerType: string.isRequired,
    /**
     * The provisioner ID associated with the runs' task.
     */
    provisionerId: string.isRequired,
    /**
     * The current selected run index to display in the card. Paging through
     * runs will trigger a history change, for which the `selectedRunId` can be
     * updated.
     */
    selectedRunId: number.isRequired,
    /**
     * Execute a function to load new artifacts when paging through them.
     */
    onArtifactsPageChange: func.isRequired,
  };

  state = {
    showArtifacts: false,
    showMore: false,
  };

  getCurrentRun() {
    return this.props.runs[this.props.selectedRunId];
  }

  isLiveLog = () => {
    const { state } = this.getCurrentRun();

    return state === 'PENDING' || state === 'RUNNING';
  };

  getArtifactUrl = ({ url, isLog }) => {
    if (!url) {
      return '';
    }

    const { taskId, runId } = this.getCurrentRun();

    if (isLog) {
      const encoded = encodeURIComponent(url);

      return this.isLiveLog()
        ? `/tasks/${taskId}/runs/${runId}/logs/live/${encoded}`
        : `/tasks/${taskId}/runs/${runId}/logs/${encoded}`;
    }

    return url;
  };

  handleNext = () => {
    const { history } = this.props;
    const { taskId, runId } = this.getCurrentRun();

    history.push(`/tasks/${taskId}/runs/${runId + 1}`);
  };

  handlePrevious = () => {
    const { history } = this.props;
    const { taskId, runId } = this.getCurrentRun();

    history.push(`/tasks/${taskId}/runs/${runId - 1}`);
  };

  handleToggleArtifacts = () => {
    this.setState({ showArtifacts: !this.state.showArtifacts });
  };

  getLiveLogArtifactFromRun = run => {
    const artifact =
      run &&
      run.artifacts.edges.find(
        ({ node: { name } }) => name === 'public/logs/live.log'
      );

    if (!artifact) {
      return;
    }

    return artifact.node;
  };

  createSortedArtifactsConnection(artifacts) {
    return {
      ...artifacts,
      edges: [...artifacts.edges].sort((a, b) => {
        if (a.node.isPublic === b.node.isPublic) {
          return 0;
        }

        return a.node.isPublic ? -1 : 1;
      }),
    };
  }

  renderArtifactsTable() {
    const { classes, onArtifactsPageChange } = this.props;
    const run = this.getCurrentRun();
    const artifacts =
      run && this.createSortedArtifactsConnection(run.artifacts);

    return (
      <ConnectionDataTable
        connection={artifacts}
        pageSize={ARTIFACTS_PAGE_SIZE}
        columnsSize={3}
        onPageChange={onArtifactsPageChange}
        withoutTopPagination
        renderRow={({ node: artifact }) => (
          <TableRow
            key={`run-artifact-${run.taskId}-${run.runId}-${artifact.name}`}
            className={classNames(
              classes.listItemButton,
              classes.artifactTableRow,
              {
                [classes.pointer]: !!artifact.url,
              }
            )}
            hover={!!artifact.url}>
            <TableCell>
              <Link
                className={classes.artifactLink}
                to={this.getArtifactUrl(artifact)}>
                <div>
                  {artifact.isPublic && <LockOpenOutlineIcon />}
                  {!artifact.isPublic && artifact.url && <LockIcon />}
                </div>
                <div className={classes.artifactNameWrapper}>
                  {artifact.isLog && (
                    <Label status="info" mini className={classes.logButton}>
                      LOG
                    </Label>
                  )}
                  <div className={classes.artifactName}>
                    <Typography>{artifact.name}</Typography>
                  </div>
                </div>
                <div>
                  {artifact.isPublic && <LinkIcon />}
                  {!artifact.isPublic && artifact.url && <OpenInNewIcon />}
                </div>
              </Link>
            </TableCell>
          </TableRow>
        )}
      />
    );
  }

  handleToggleMore = () => {
    this.setState({ showMore: !this.state.showMore });
  };

  render() {
    const {
      classes,
      runs,
      selectedRunId,
      provisionerId,
      workerType,
      theme,
    } = this.props;
    const { showArtifacts, showMore } = this.state;
    const run = this.getCurrentRun();
    const liveLogArtifact = this.getLiveLogArtifactFromRun(run);

    return (
      <Card raised>
        <div>
          <CardContent classes={{ root: classes.cardContent }}>
            <Typography variant="h5" className={classes.headline}>
              {run ? `Task Run ${selectedRunId}` : 'Task Run'}
            </Typography>
            {run ? (
              <Fragment>
                <List>
                  <ListItem>
                    <ListItemText
                      primary="State"
                      secondary={<StatusLabel state={run.state} />}
                    />
                  </ListItem>
                  {liveLogArtifact && (
                    <ListItem
                      button
                      className={classes.listItemButton}
                      component={Link}
                      to={this.getArtifactUrl(liveLogArtifact)}>
                      <ListItemText
                        primary={
                          <Fragment>
                            View Live Log{' '}
                            <Label
                              status="info"
                              mini
                              className={classes.liveLogLabel}>
                              LOG
                            </Label>
                          </Fragment>
                        }
                        secondary={liveLogArtifact.name}
                      />
                      <LinkIcon />
                    </ListItem>
                  )}
                  <ListItem
                    button
                    className={classes.listItemButton}
                    onClick={this.handleToggleArtifacts}>
                    <ListItemText primary="Artifacts" />
                    {showArtifacts ? <ChevronUpIcon /> : <ChevronDownIcon />}
                  </ListItem>
                  <Collapse in={showArtifacts} timeout="auto">
                    <List component="div" disablePadding>
                      <ListItem
                        className={classes.artifactsListItemContainer}
                        component="div"
                        disableGutters>
                        {this.renderArtifactsTable()}
                        <Divider />
                      </ListItem>
                    </List>
                  </Collapse>
                  <CopyToClipboard
                    title={`${run.scheduled} (Copy)`}
                    text={run.scheduled}>
                    <ListItem button className={classes.listItemButton}>
                      <ListItemText
                        primary="Scheduled"
                        secondary={<DateDistance from={run.scheduled} />}
                      />
                      <ContentCopyIcon />
                    </ListItem>
                  </CopyToClipboard>
                  <CopyToClipboard
                    title={`${run.started} (Copy)`}
                    text={run.started}>
                    <ListItem button className={classes.listItemButton}>
                      <ListItemText
                        primary="Started"
                        secondary={
                          run.started ? (
                            <DateDistance
                              from={run.started}
                              offset={run.scheduled}
                            />
                          ) : (
                            <em>n/a</em>
                          )
                        }
                      />
                      <ContentCopyIcon />
                    </ListItem>
                  </CopyToClipboard>
                  <CopyToClipboard
                    title={`${run.resolved} (Copy)`}
                    text={run.resolved}>
                    <ListItem button className={classes.listItemButton}>
                      <ListItemText
                        primary="Resolved"
                        secondary={
                          run.resolved ? (
                            <DateDistance
                              from={run.resolved}
                              offset={run.started}
                            />
                          ) : (
                            <em>n/a</em>
                          )
                        }
                      />
                      <ContentCopyIcon />
                    </ListItem>
                  </CopyToClipboard>
                  <ListItem
                    button
                    className={classes.listItemButton}
                    onClick={this.handleToggleMore}>
                    <ListItemText primary={showMore ? 'Less...' : 'More...'} />
                    {showMore ? <ChevronUpIcon /> : <ChevronDownIcon />}
                  </ListItem>
                </List>
                <Collapse in={showMore} timeout="auto">
                  <List component="div" disablePadding>
                    <ListItem>
                      <ListItemText
                        primary="Reason Created"
                        secondary={<StatusLabel state={run.reasonCreated} />}
                      />
                    </ListItem>
                    <ListItem>
                      <ListItemText
                        primary="Reason Resolved"
                        secondary={
                          run.reasonResolved ? (
                            <StatusLabel state={run.reasonResolved} />
                          ) : (
                            <em>n/a</em>
                          )
                        }
                      />
                    </ListItem>
                    <ListItem>
                      <ListItemText
                        primary="Worker Group"
                        secondary={run.workerGroup || <em>n/a</em>}
                      />
                    </ListItem>
                    <ListItem
                      title="View Worker"
                      button
                      className={classes.listItemButton}
                      component={Link}
                      to={`/provisioners/${provisionerId}/worker-types/${workerType}/workers/${run.workerId}`}>
                      <ListItemText
                        primary="Worker ID"
                        secondary={run.workerId}
                      />
                      <LinkIcon />
                    </ListItem>
                    <CopyToClipboard
                      title={`${run.takenUntil} (Copy)`}
                      text={run.takenUntil}>
                      <ListItem button className={classes.listItemButton}>
                        <ListItemText
                          primary="Taken Until"
                          secondary={
                            run.takenUntil ? (
                              <DateDistance from={run.takenUntil} />
                            ) : (
                              <em>n/a</em>
                            )
                          }
                        />
                        <ContentCopyIcon />
                      </ListItem>
                    </CopyToClipboard>
                  </List>
                </Collapse>
              </Fragment>
            ) : (
              <div className={classes.boxVariant}>
                <NoRunsIcon
                  fill={theme.palette.text.primary}
                  className={classes.boxVariantIcon}
                />
                <Typography className={classes.boxVariantText} variant="h6">
                  No Runs
                </Typography>
                <Typography className={classes.boxVariantText}>
                  A run will be created when the task gets schedueled.
                </Typography>
              </div>
            )}
          </CardContent>
          <MobileStepper
            variant={runs.length > DOTS_VARIANT_LIMIT ? 'progress' : 'dots'}
            position="static"
            steps={runs.length}
            activeStep={selectedRunId}
            className={classes.root}
            nextButton={
              <Button
                size="small"
                onClick={this.handleNext}
                disabled={run ? selectedRunId === runs.length - 1 : true}>
                Next
                <ChevronRightIcon />
              </Button>
            }
            backButton={
              <Button
                size="small"
                onClick={this.handlePrevious}
                disabled={run ? selectedRunId === 0 : true}>
                <ChevronLeftIcon />
                Previous
              </Button>
            }
          />
        </div>
      </Card>
    );
  }
}
