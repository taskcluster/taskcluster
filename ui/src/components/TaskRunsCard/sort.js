export const getPriority = a => {
  if (a?.node?.name?.includes('live.log')) {
    return 1;
  }

  if (a?.node?.name?.includes('live_backing.log')) {
    return 2;
  }

  return a?.node?.name?.startsWith('public/') ? 3 : 4;
};

export const sortArtifacts = artifacts => {
  return artifacts
    .map(a => ({ ...a, priority: getPriority(a) }))
    .sort((a, b) => {
      if (a.priority === b.priority) {
        return a.node?.name?.localeCompare(b.node?.name);
      }

      return a.priority - b.priority;
    });
};
