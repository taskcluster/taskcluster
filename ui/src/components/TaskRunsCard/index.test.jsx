import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import TaskRunsCard from './index';

it('should render TaskRunsCard', () => {
  const { asFragment } = render(
    <MemoryRouter keyLength={0}>
      <TaskRunsCard
        taskQueueId="task/queueId"
        task={{
          taskId: 'taskId',
          status: {
            taskId: 'taskId',
            provisioinerId: 'p1',
            workerType: 'workerType',
            taskGroupId: 'tg1',
            state: 'RUNNING',
          },
          taskQueueId: 'task/queueId',
          created: new Date().toISOString(),
          deadline: new Date().toISOString(),
          expires: new Date().toISOString(),
          scopes: ['scopes'],
          routes: [],
          payload: {
            command: [
              '/bin/bash',
              '-c',
              'for ((i=1;i<=600;i++)); do echo $i; sleep 1; done',
            ],
            image: 'ubuntu:latest',
            maxRunTime: 630,
          },
          metadata: {
            source: 'https://test.taskcluster-dev.net/task/taskId',
          },
          extra: {},
        }}
        selectedRunId=""
        runs={[
          {
            taskId: 'eR1kMya2SruyMaRMZguROg',
            runId: 0,
            state: 'COMPLETED',
            reasonCreated: 'SCHEDULED',
            reasonResolved: 'COMPLETED',
            scheduled: '2022-02-03T14:41:19.706Z',
            started: '2022-02-03T14:43:54.086Z',
            resolved: '2022-02-03T14:45:28.396Z',
            workerGroup: 'us-east1',
            workerId: '7421215367664916236',
            takenUntil: '2022-02-03T15:03:54.082Z',
            artifacts: {
              pageInfo: {
                hasNextPage: false,
                hasPreviousPage: false,
                cursor: '$$FIRST$$',
                previousCursor: null,
                nextCursor: null,
                __typename: 'PageInfo',
              },
              edges: [
                {
                  node: {
                    name: 'public/coverage-final.json',
                    contentType: 'application/json',
                    __typename: 'Artifact',
                  },
                  __typename: 'ArtifactsEdge',
                },
                {
                  node: {
                    name: 'public/logs/live_backing.log',
                    contentType: 'text/plain; charset=utf-8',
                    __typename: 'Artifact',
                  },
                  __typename: 'ArtifactsEdge',
                },
                {
                  node: {
                    name: 'public/logs/live.log',
                    contentType: 'text/plain; charset=utf-8',
                    __typename: 'Artifact',
                  },
                  __typename: 'ArtifactsEdge',
                },
              ],
              __typename: 'ArtifactsConnection',
            },
            __typename: 'TaskRun',
          },
        ]}
      />
    </MemoryRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
