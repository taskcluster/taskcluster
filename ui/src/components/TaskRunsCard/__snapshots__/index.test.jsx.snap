// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render TaskRunsCard 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation8 MuiPaper-rounded"
  >
    <div>
      <div
        class="MuiCardContent-root TaskRunsCard-cardContent-2"
      >
        <div
          class="MuiPaper-root MuiMobileStepper-root MuiMobileStepper-positionStatic MuiPaper-elevation0"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text TaskRunsCard-previousPageArrow-16 MuiButton-textSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              <svg
                class="mdi-icon "
                fill="currentColor"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z"
                />
              </svg>
              Previous
            </span>
          </button>
          <div
            class="MuiMobileStepper-dots"
          >
            <div
              class="MuiMobileStepper-dot"
            />
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text TaskRunsCard-nextPageArrow-17 MuiButton-textSizeSmall MuiButton-sizeSmall Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Next
              <svg
                class="mdi-icon "
                fill="currentColor"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M8.59,16.58L13.17,12L8.59,7.41L10,6L16,12L10,18L8.59,16.58Z"
                />
              </svg>
            </span>
          </button>
        </div>
        <h5
          class="MuiTypography-root TaskRunsCard-headline-1 MuiTypography-h5"
        >
          Task Run
        </h5>
        <div
          class="TaskRunsCard-boxVariant-9"
        >
          <svg
            class="TaskRunsCard-boxVariantIcon-8"
            height="24px"
            version="1.1"
            viewBox="0 0 24 24"
            width="24px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2,10.96C1.5,10.68 1.35,10.07 1.63,9.59L3.13,7C3.24,6.8 3.41,6.66 3.6,6.58L11.43,2.18C11.59,2.06 11.79,2 12,2C12.21,2 12.41,2.06 12.57,2.18L20.47,6.62C20.66,6.72 20.82,6.88 20.91,7.08L22.36,9.6C22.64,10.08 22.47,10.69 22,10.96L21,11.54V16.5C21,16.88 20.79,17.21 20.47,17.38L12.57,21.82C12.41,21.94 12.21,22 12,22C11.79,22 11.59,21.94 11.43,21.82L3.53,17.38C3.21,17.21 3,16.88 3,16.5V10.96C2.7,11.13 2.32,11.14 2,10.96M12,4.15V4.15L12,10.85V10.85L17.96,7.5L12,4.15M5,15.91L11,19.29V12.58L5,9.21V15.91M19,15.91V12.69L14,15.59C13.67,15.77 13.3,15.76 13,15.6V19.29L19,15.91M13.85,13.36L20.13,9.73L19.55,8.72L13.27,12.35L13.85,13.36Z"
              fill="rgba(0, 0, 0, 0.87)"
              opacity="0.05"
            />
          </svg>
          <h6
            class="MuiTypography-root TaskRunsCard-boxVariantText-10 MuiTypography-h6"
          >
            No Runs
          </h6>
          <p
            class="MuiTypography-root TaskRunsCard-boxVariantText-10 MuiTypography-body2"
          >
            A run will be created when the task gets scheduled.
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
